src/group/stabchain/builder/random.rs

```diff
@@ -0,0 +1,435 @@
+use crate::group::orbit::abstraction::FactoredTransversalResolver;
+use crate::group::orbit::transversal::factored_transversal::{
+    factored_transversal_complete_opt, representative_raw_as_word,
+};
+use crate::group::stabchain::element_testing::residue_as_words_from_words;
+use crate::group::stabchain::{MovedPointSelector, Stabchain, StabchainRecord};
+use crate::group::utils::{
+    apply_permutation_word, collapse_perm_word, random_subproduct_word_full,
+    random_subproduct_word_subset,
+};
+use crate::group::Group;
+use crate::perm::actions::SimpleApplication;
+use crate::perm::{Action, Permutation};
+use itertools::Itertools;
+use rand::rngs::ThreadRng;
+use rand::seq::IteratorRandom;
+use rand::thread_rng;
+use std::collections::{HashMap, VecDeque};
+use std::iter::Iterator;
+use std::iter::{repeat_with, FromIterator};
+
+//Constants for subproduct generation
+const C1: usize = 10;
+const C2: usize = 10;
+const ORBIT_BOUND: usize = 50;
+const BASE_BOUND: usize = 5;
+// Cosntants for the Strong Generating Test.
+const C3: usize = 10;
+const C4: usize = 10;
+
+// Helper struct, used to build the stabilizer chain
+
+pub struct StabchainBuilderRandom<P, S, A = SimpleApplication<P>>
+where
+    A: Action<P, OrbitT = usize>,
+{
+    current_pos: usize,
+    chain: Vec<StabchainRecord<P, FactoredTransversalResolver<A>, A>>,
+    selector: S,
+    action: A,
+    n: usize,
+    base: Vec<A::OrbitT>,
+    rng: ThreadRng,
+    //The chain is zero indexed, but this field is 1 indexed.
```

Comment:
The code works by decrementing the up to date level, so it is a lot simpler and would avoid the need to check if it will become zero whenever this is modified.